src/main/java/org/folio/rest/impl/PutOrderLineByIdHelper.java

```diff
@@ -161,14 +168,91 @@ public boolean isUpdateSuccessful() {
       .thenCompose(jsonObj -> updateOrderLine(compPOL, jsonObj))
       .thenCompose(v -> inventoryHelper.getOrCreateHoldingsRecord(compPOL))
       .thenCompose(holdingsId -> inventoryHelper.handleItemRecords(compPOL, holdingsId))
-      .thenAccept(itemIds -> {
+      .thenCompose(itemIds -> {
         // Temporal check. The idea is to create piece records for successfully created items and then throw exception
         if (itemIds.size() != expectedItemsQuantity) {
           throw new IllegalStateException("Expected items quantity does not correspond to created items");
         }
-      });
+        return completedFuture(itemIds);
+      })
+      .thenCompose(itemIds -> createPieces(compPOL, itemIds));
   }
 
+  /**
+   * Creates Piece records corresponding to each item record associated with PO Line
+   *
+   * @param compPOL Composite PO line to update Inventory for
+   * @param expectedItemsQuantity Inventory items quantity
+   * @param itemIds List of item id
+   * @return CompletableFuture with newly created Pieces associated with PO line.
+   */
+  private CompletableFuture<Void> createPieces(PoLine compPOL, List<String> itemIds) {
+		CompletableFuture<Void> future = new VertxCompletableFuture<>(ctx);
+		if (itemIds.isEmpty()) {
+			future.complete(null);
+			return future;
+		}
+		List<CompletableFuture<Void>> futuresList = new ArrayList<>();
+		String poLineId = compPOL.getId();		
+		String endpoint = String.format(PIECES_ENDPOINT, poLineId);
+		
+		handleGetRequest(endpoint, httpClient, ctx, okapiHeaders, logger)
+		    .thenAccept(body -> {
+			    PieceCollection pieces = body.mapTo(PieceCollection.class);
+		    	List<String> existingItemIds = pieces.getPieces()
+							.stream()
+							.map(Piece::getItemId)
+							.filter(StringUtils::isNotEmpty)
+							.collect(Collectors.toList());
+			    	
+				 itemIds.stream()
+				 				.filter(id -> !existingItemIds.contains(id))
+				 				.forEach(itemId -> futuresList.add(createPiece(poLineId, itemId)));
+
+			    logger.info("response from GET /pieces: " + body.encodePrettily());
+
+			    CompletableFuture.allOf(futuresList.toArray(new CompletableFuture[0]))
+			    .thenAccept(v -> future.complete(null))
+			    .exceptionally(t -> {
+				    future.completeExceptionally(t);
+				    return null;
+			    });
+		    });
```

Comment:
Only more note. The `future` is being created in the beginning and expected to be completed manually. In case `handleGetRequest` completes exceptionally, the `future` will never be completed